sdk/python/lib/pulumi/x/automation/_stack.py

```diff
@@ -620,3 +670,45 @@ def fully_qualified_stack_name(org: str, project: str, stack: str) -> str:
     :returns: The fully qualified stack name.
     """
     return f"{org}/{project}/{stack}"
+
+
+def _create_log_file(command: str) -> Tuple[str, tempfile.TemporaryDirectory]:
+    log_dir = tempfile.TemporaryDirectory(prefix=f"automation-logs-{command}-")
+    filepath = os.path.join(log_dir.name, "eventlog.txt")
+
+    # Open and close the file to ensure it exists before we start polling for logs
+    f = open(filepath, "w+")
+    f.close()
+    return filepath, log_dir
+
+
+def _watch_logs(filename: str, callback: OnEvent):
+    with open(filename) as f:
+        while True:
+            line = f.readline()
+
+            # sleep if file hasn't been updated
+            if not line:
+                time.sleep(0.1)
+                continue
+
+            event = EngineEvent.from_json(json.loads(line))
+            callback(event)
+
+            # if this is the cancel event, stop watching logs.
```

Comment:
I'd really like to know more about this one, what kind of cancel event? I do not think I have matching logic in dotnet SDK, perhaps a miss.

Reply:
`CancelEvent` is the last event transmitted by the engine as a signal that logging is over. I've used it here to signal that the thread can terminate, so when `thread.join()` waits it doesn't always time out because otherwise there would be no end to this `_watch_logs` function. We can also use an `Event` object to [signal completion](https://docs.python.org/3/library/threading.html#id1:~:text=A%20thread%20can%20be%20flagged%20as,signalling%20mechanism%20such%20as%20an%20Event.) though, if there is hesitation to go this route.

Reply:
```// CancelEvent is emitted when the user initiates a cancellation of the update in progress, or
// the update successfully completes.
export interface CancelEvent {
}
```

So it is not the above Cancel event? And nothing about stack.Cancel() right?

Yeah this is awesome that the engine signals completion of logging. I wanted something like his for .NET to avoid race conditions, but I did not realize there is an event we can count on to be the last event. Calling it "cancel event" seems a little misleading but it does seem like we need something like this.

If we use built-in concurrency primitives to try to cancel the bg thread when we think the work is done, we're always open to a subtle race condition where we guess wrong (we think the work is done but IO has not propagated all the logs yet).

Reply:
It is indeed the same `CancelEvent` that you referenced. It is emitted on `stack.cancel()` but also whenever an update completes, in both the error and success case. It's not so much a "completion of logging" but a "completion of update" which I am equating to the same thing here. I admit I'm not 100% confident that this will always be the case, but we do have the timeout on `thread.join()` to fall back on in any case (maybe that should be shorter?).

> Calling it "cancel event" seems a little misleading but it does seem like we need something like this.

Heh, agreed 😄 

I'm tempted to go ahead with this and we can make adjustments if we run into issues. All of our error tests complete successfully so I'm not too worried.

Reply:
Yes please, don't block on this. I will follow up to understand this better. Perhaps we can rename things later to make more sense.